chore(memory): promote strong-fit contacts drafts for review (#132)
medic/cht-agent: agent-memory/domains/contacts/issues/9835-cht-datasource-create-update-contacts.md

## Root Cause

The initial `cht-datasource` implementation focused on reads. Write operations were scattered across different services without a unified interface: validation was repeated at each call site, currying patterns were inconsistent, writes cost extra DB round-trips, and error types were used with the wrong semantics. Those are the conditions the create/update surface below was written to replace.

## Solution

**Provenance — read this before citing any PR below.** `#10083` is the *umbrella* PR that squashed the whole `9835` feature branch onto master (squash `f382785be`, 2026-03-10). Everything described in this draft shipped in that one commit; master had no cht-datasource write support at all beforehand. `#10522` (squash `a89955a9f`, 2026-03-03) is an **ancestor of `#10083`'s head branch**, not a follow-up to it: `#10522` refactored the create/update surface that earlier child PRs had built on the branch, and then the branch landed as `#10083`. Every other PR cited here (`#10022`, `#10081`, `#10222`, `#10246`) likewise merged into the feature branch, not into master. Do not read the PR numbers as a chronology of master.

> The create/update surface was built incrementally on the feature branch: person/place/report create and update for both the local (PouchDB) and remote (HTTP) data contexts, the `input.ts` module, and new endpoints wired through the person/place/report api controllers, `routing.js`, `auth.js`, and `server-utils.js`. The initial local/API `createPerson` work arrived via branch PRs that cite sub-issue `#10036` rather than `#9835`, so they are deliberately not listed in `source_prs` here. The contact controller was modified as well but received **no** write endpoint — it was only migrated to `auth.assertPermissions`; on master it still exposes only `get`, `getUuids`, `getAll`, `getSummaries`, and `routing.js` gained no contact write route.

Reviewer: **suggestion (non-blocking):** "it was **only** migrated to `auth.assertPermissions`" undersells the
same commit it cites: `f382785be` also reshaped contact.js's datasource wiring, splitting the
per-request `ctx.bind(with_lineage === 'true' ? ... : ...)` into two eagerly bound getters. The
no-write-endpoint half is exactly right.

```suggestion
The create/update surface was built incrementally on the feature branch: person/place/report create and update for both the local (PouchDB) and remote (HTTP) data contexts, the `input.ts` module, and new endpoints wired through the person/place/report api controllers, `routing.js`, `auth.js`, and `server-utils.js`. The initial local/API `createPerson` work arrived via branch PRs that cite sub-issue `#10036` rather than `#9835`, so they are deliberately not listed in `source_prs` here. The contact controller was modified as well but received **no** write endpoint — it was migrated to `auth.assertPermissions` and its datasource bindings were reshaped (`getContact`/`getContactWithLineage` bound eagerly instead of per-request); on master it still exposes only `get`, `getUuids`, `getAll`, `getSummaries`, and `routing.js` gained no contact write route.
```

- Flattened `isPlace/isPerson/isContact` from curried `(settings) => (doc, uuid?)` to direct two-argument `(settings, doc)` form. This applies to the type guards only — the db-bound helpers (`getDocsByIds`, `createDoc`, `updateDoc`, `minifyDoc`) all remain curried on the PouchDB handle.
- `getDocsByIds` now returns `Nullable<Doc>[]` preserving index positions for parallel lookups; it previously de-duplicated and filtered down to `Doc[]`, so positions were not stable
- `createDoc/updateDoc` no longer re-fetch the document after writing (both previously ended in `getDocById(db)(id)`)
- Lineage handling centralized in `local/libs/lineage.ts` with `minifyDoc`, `assertSameParentLineage`, `getUpdatedContact`. The file itself predates this work; `#10522` consolidated write-side lineage logic into it.
- Input types cleaned up — `_id` and `_rev` are `never` on create inputs (`ContactInput`, `ReportInput`, `PlaceInput`; `PersonInput extends ContactInput`)

> - Write-input validation moved into `src/libs/parameter-validators.ts` as composable assertion functions — `assertPersonInput`/`assertPlaceInput`/`assertReportInput`, all built on a shared `assertContactInput`. `src/input.ts` was left as a **types-only** module. Earlier on the branch `input.ts` did hold the validators (the controllers called `Input.validatePersonInput(req.body)`), which is why pre-`#10522` history reads differently — see Design Choices.

Reviewer: **suggestion (non-blocking):** "all built on a shared `assertContactInput`" is true for two of the
three. `assertPersonInput` (`parameter-validators.ts:69`) and `assertPlaceInput` (`:83`) each open with
`assertContactInput(data)`; `assertReportInput` (`:92-103`) never calls it, repeating the
object/reported_date/no-`_id`-or-`_rev` checks itself against `form` and `contact` (ReportInput has no
`name`, so it cannot reuse the contact assertion). Asserted reuse where none exists is the shape this
draft's own Code Patterns section warns against.

```suggestion
- Write-input validation moved into `src/libs/parameter-validators.ts` as composable assertion functions — `assertPersonInput`/`assertPlaceInput`, both built on a shared `assertContactInput`, plus a standalone `assertReportInput` that repeats the same object/reported_date/no-`_id`-or-`_rev` checks against `form` and `contact` instead (ReportInput has no `name`, so it cannot reuse the contact assertion). `src/input.ts` was left as a **types-only** module. Earlier on the branch `input.ts` did hold the validators (the controllers called `Input.validatePersonInput(req.body)`), which is why pre-`#10522` history reads differently — see Design Choices.
```

> 

Reviewer: issue (blocking) [agent-memory/domains/contacts/issues/9835-cht-datasource-create-update-contacts.md:76]: The permission enumeration is wrong for five of the six write endpoints. At the #10083 merge commit (f382785be) and unchanged on current master (6eb2ac0f1), the checks are: person create hasAny ['can_create_people', 'can_edit'] (person.js:152), person update hasAny ['can_update_people', 'can_edit'] (person.js:202), place create hasAny ['can_create_places', 'can_edit'] (place.js:151), place update hasAny ['can_update_places', 'can_edit'] (place.js:209), report create hasAny ['can_create_records', 'can_edit'] (report.js:268), report update hasAny ['can_update_reports', 'can_edit'] (report.js:326), all with isOnline: true. The removed read permission was also can_view_reports for report endpoints, not can_view_contacts. Suggested replacement: "Permission fix (PR #10522): write endpoints no longer require the read permission (can_view_contacts for person/place, can_view_reports for report); each requires an online user with can_edit or the entity-specific write permission (can_create_people/can_update_people, can_create_places/can_update_places, can_create_records/can_update_reports)." Relatedly, the Solution paragraph's claim that PR #10222 made person/place/report honor can_edit is imprecise: #10222's diff adds can_edit for person and place only; report gained can_edit in #10522, which also fixed report update's constant from can_update_records to can_update_reports.

suggestion (non-blocking) [agent-memory/domains/contacts/issues/9835-cht-datasource-create-update-contacts.md:11]: The Solution section leads with PR #10522 (feat(#9835): refactor create/update cht-datasource apis, merged into the feature branch 2026-03-03) and describes its changes in detail, but source_prs omits it. Add "medic/cht-core#10522" to source_prs.

(Nitpick about the Domain note's "its anchor PR" wording: drop. The draft has an identifiable singular anchor PR, #10083, the only source PR merged to master; the claim attached to it is true, and the identical boilerplate appears in sibling multi-PR drafts 10043 and 10057.)

(Anchored here because line 76, where the enumeration lives, is outside this round's diff hunks.)

- Qualifier-based data access: `src/qualifier.ts` builds the read/lookup qualifiers (`Qualifier.byUuid`, `byContactType`, `byFreetext`) that the local and remote data sources consume. PR #10081 reshaped its create/update qualifiers only — it did not change how the person data source reads.

## Design Choices

- `ResourceNotFoundError` semantically distinguishes "document not found" from "bad argument" — mapped to HTTP 404 in `api/src/server-utils.js`, and back to the error class from a 404 response in the remote adapter (`remote/libs/data-context.ts`)
- Index-preserving `getDocsByIds` enables parallel lookups. Note the currying: `const getMedicDocsByIds = getDocsByIds(medicDb); const [parent, contact] = await getMedicDocsByIds([parentId, contactId]);` — `getDocsByIds`, `createDoc`, `updateDoc` and `minifyDoc` are all `(db) => (args) => ...`, so the db handle is bound once at the top of each operation
- Lineage minification reuses the `minify` function from the `@medic/lineage` package (`shared-libs/lineage/src/minify.js`) rather than reimplementing dehydration

> - Permission fix: the write endpoints no longer require the matching **read** permission (`can_view_contacts` for person/place, `can_view_reports` for report). Each now asserts `isOnline: true` plus `hasAny` of its own write permission or the general `can_edit`. The six write endpoints do **not** share one permission pair — per endpoint on master:

Reviewer: **praise:** The blocking permission finding was not just patched but replaced with a per-endpoint table
that is exact on a master that has moved since the round-3 review, and the surrounding provenance
rewrite verifies claim by claim. Same standard elsewhere in this delta: the 10344 epic-branch banner
verifies end to end including every comment in its shell block, and the 10057 `#10124` attribution is
now more precise than the round-3 comment that asked for it (the helpers are closures inside the create
factories, not module-level). The 8684 config/standard exception and the 9090 importer recount also
reproduce exactly. This is the strongest repair round of the four memory PRs.

## Related Files

- shared-libs/cht-datasource/src/index.ts
- shared-libs/cht-datasource/src/input.ts
- shared-libs/cht-datasource/src/qualifier.ts (PR #10081)
- shared-libs/cht-datasource/src/local/person.ts
- shared-libs/cht-datasource/src/local/place.ts
- shared-libs/cht-datasource/src/local/report.ts
- shared-libs/cht-datasource/src/remote/person.ts (PR #10083)
- shared-libs/cht-datasource/src/remote/report.ts (PR #10083)
- shared-libs/cht-datasource/src/local/libs/lineage.ts
- shared-libs/cht-datasource/src/local/libs/doc.ts
- shared-libs/cht-datasource/src/local/libs/core.ts (`assertFieldsUnchanged`, `getReportedDateTimestamp`; hosted the `ensureHasRequiredFields` helper patched by PR #10246)
- shared-libs/cht-datasource/src/libs/parameter-validators.ts
- shared-libs/cht-datasource/src/libs/error.ts
- api/src/auth.js
- api/src/controllers/contact.js (PR #10083)
- api/src/controllers/person.js
- api/src/controllers/place.js
- api/src/controllers/report.js (PR #10083)
- api/src/routing.js (PR #10083)
- api/src/server-utils.js

## Testing

> - Unit tests updated across most changed cht-datasource modules — 19 spec files under `shared-libs/cht-datasource/test/` in the `#10083` landing, against 23 changed `src` files. Not every changed module got a spec: `src/input.ts` was added with no spec at all (there is no `test/input.spec.ts` on master), and `src/libs/error.ts`, `src/libs/doc.ts` and `src/local/libs/data-context.ts` changed without a matching spec change in that commit.

Reviewer: **nitpick (non-blocking):** The spec-gap list reads as exhaustive but omits one: `src/libs/constants.ts`
also changed in `f382785be` with no matching spec, and no constants spec exists anywhere under
`shared-libs/cht-datasource/test/` on master.

```suggestion
- Unit tests updated across most changed cht-datasource modules — 19 spec files under `shared-libs/cht-datasource/test/` in the `#10083` landing, against 23 changed `src` files. Not every changed module got a spec: `src/input.ts` was added with no spec at all (there is no `test/input.spec.ts` on master), and `src/libs/error.ts`, `src/libs/doc.ts`, `src/libs/constants.ts` and `src/local/libs/data-context.ts` changed without a matching spec change in that commit.
```
